### ISSUE TITLE
How to use beforeRequest to override query

### ISSUE THREAD

**Issue by tahayk** [ptejada/artillery-engine-socketio-v3#8]
Is it possible to override `query` in `beforeRequest` ? 
I want to provide a dynamic value in query, that changes between each request but I couldn't find a way to do it.
Any idea how to do it ?

**Comment by mehdifracso-ilearn** [comment]
hey @tahayk, did you find an answer?

**Comment by ptejada** [comment]
@mehdifracso-ilearn 
Today it is only possible to configure query params at the engine config level
![image](https://user-images.githubusercontent.com/253405/193422583-92d2b089-dd2e-4931-a4a2-9903ae5addba.png)

However, we can add the functionality by duplicating what we currently do for the `extraHeaders` option.

**Comment by gnom1gnom** [comment]
Can you please elaborate how to properly set the extraHeaders within a script.
I implemented socket auth as described in https://socket.io/docs/v4/middlewares/#sending-credentials.
The authorization middleware is called immediately when you connect to a socket, even before the actual artillery scenario is executed.
I injected extraHeaders to the context in a before section
```
config:
  target: "http://localhost:4000"
  phases:
    - duration: 1
      arrivalRate: 1
  engines:
    socketio-v3: {}
  processor: "./artillery-processor.js"
before:
    flow:
      - get:
          url: "/api/auth/XXXXXXXXX"
          afterResponse: "storeToken"
      - get:
          url: "/api/health"
          beforeRequest: "logToken"
scenarios:
  - name: Test chat
    engine: socketio-v3
    flow:
```
The processor script looks as follows:
```
/* eslint-disable no-unused-vars */
'use strict';

const jwtRe = /(?<=token=)[\w-]*\.[\w-]*\.[\w-]*(?=;)/

module.exports = {
	storeToken: (requestParams, response, context, ee, next) => {
		const token = jwtRe.exec(...response.headers["set-cookie"])[0];
		context.extraHeaders = { auth: { token: token } };
		return next();
	},
	logToken: (requestParams, context, ee, next) => {
		console.log(context.extraHeaders.auth.token);
		return next();
	}
};
```
The log token function indicates that the token has been successfully extracted and stored in context.extraHeaders.

I modified the /node_modules/artillery-engine-socketio-v3/index.js to see if the context contains the extraHeaders. Unfortunately they are undefined.
```
    console.log(`context.extraHeaders`, context.extraHeaders);
    const socketioOpts = template(
      Object.assign({}, this.socketioOpts, {
        extraHeaders: {
          ...this.socketioOpts.extraHeaders,
          ...context.extraHeaders
        },
```
```
Phase started: unnamed (index: 0, duration: 1s) 17:26:15(+0000)
⠋ context.extraHeaders undefined
```

**Comment by ptejada** [comment]
I am unsure how the `before` step works or whether the context is shared across engines. Unlike the example provided [here ](https://github.com/ptejada/artillery-engine-socketio-v3#testing-socket-connection-and-re-connections) you are using the HTTP engine to get the auth token instead of the using the socket.io one.

I would check how you can capture data from the `before` stage in the steps of the main scenarios.

**Comment by jeffwillden** [comment]
When a socketio-v3 test runs in artillery, it initiates the socket.io connection **before** any HTTP GET requests or POST requests. Because of this behavior, the following is not possible:

```
engines:
  socketio-v3:
    query:
      token: "{{ jwtFromPOST }}"
```
 The variable is not yet populated when artillery initiates the socket connections.

This also means that socketio-v3 tests differ somewhat from the way a real browser would work. The browser must first GET the javascript `/socket.io/socket.io.js` and then it can initiate a socket.io connection. Artillery tests using socketio-v3 must do it the other way around.

Maybe it would make sense to delay initiating the socket.io connection until it is invoked in a scenario. When that happens, {{ variables }} could already be populated, and the expected behavior (with a jwt like above) would work properly.

**Background**
socketio-v3 works fine if authentication occurs over the socket. Some examples on the net show `emit('login', {user:'name', pass:'word'})` but this of course requires JavaScript to **know** the credentials. There are good reasons to **never** pass credentials around in JavaScript, but only have the user send them in a POST request. In that case, the server responds with a JSON web token (JWT) which is required upon initiating a socket connection. If the token is missing or invalid, the socket is closed immediately. This design helps reduce attack vectors and vulnerabilities. Currently socketio-v3 only supports designs that expose more potential vulnerabilities.

You mention above using `before` to capture from a POST, but in my tests, the socketio connection is initiated before the `before` step runs, so that might not be a viable alternative.

**Comment by gnom1gnom** [comment]
I managed to fix this isue and created a PR #11.
The trick was to:
1.  Store the token in the `context.vars` as it is not lost between HTTP and Socket engine executions
#### **`scenario.yml`**
``` yml
config:
  target: "http://localhost:4000"
  processor: "./artillery-processor.js"
  engines:
    socketio-v3: {}
before:
  flow:
    - get:
       url: "/api/auth"
       afterResponse: "storeToken"
```
#### **`artillery-processor.js`**
``` js
const jwtRe = /(?<=token=)[\w-]*\.[\w-]*\.[\w-]*(?=;)/

module.exports = {
    storeToken: (requestParams, response, context, ee, next) => {
        const token = jwtRe.exec(...response.headers["set-cookie"])[0];
        context.vars.auth = { token: token };
        return next();
    }
};
```
2. Add the dedicated `auth` field in the `socketioOpts` initialization
#### **`artillery-engine-socketio-v3/index.js`**
``` js
    const socketioOpts = template(
      Object.assign({}, this.socketioOpts, {
        extraHeaders: {
          ...this.socketioOpts.extraHeaders,
          ...context.extraHeaders
        },
        auth: {
          ...context.vars.auth
        },
        // Require to force socket.io-client to set new headers on connection to different Namespace
        forceNew: true,
      }),
      context
    );
```

**Comment by jeffwillden** [comment]
Thank you for posting this. I agree with adding `auth` to `socketioOpts` in order to support Socket.io as described in the documentation. However, I do not agree with storing the value in `context.vars.auth` during the `before` stage. The `before` declaration is not executed once per socket connection, it is only executed one time. Then every socket connection each time the scenario runs will be using the same auth header. So instead of running load tests for 50,000 different users, it would be the same user logging in 50,000 times.

That might be acceptable for some load tests, but not all. It would be best to provide an example where a different token is stored each time the scenario is run. That would allow for a load test of 50,000 different users, which is probably what people want to do most often.

**Comment by ptejada** [comment]
I pushed a few updates to give more control over the Socket.IO connection. All the changes are listed in the [unreleased section of the changelog ](https://github.com/ptejada/artillery-engine-socketio-v3/blob/master/CHANGELOG.md#unreleased)

The main addition is a `connect`, which allows you to statically or dynamically configure the Socket.IO client. Ex:

```yml
...
    flow:
      - get:
          url: http://service.to/get-token
          capture:
            json: "$.token"
            as: "token"
      - connect:
          auth:
            type: uuid
            token: "{{token}}"
        namespace: /my-namespace
      - emit: [ "ping" ]
        namespace: /my-namespace
        acknowledge:
          args: true
...
```

Any `query` or `extraHeaders` settings can be configured in the same way as well.

You can test this last change by installing the package like this
```
yarn add ptejada/artillery-engine-socketio-v3

OR

npm install ptejada/artillery-engine-socketio-v3
```

Let me know if anyone is able to test these updates and confirm are working as expected.

**Comment by gnom1gnom** [comment]
I'll test it next week. Thank you for your support.

**Comment by ptejada** [comment]
I have tagged the latest release as `v1.2.0` so the latest change can be pulled normally from npm
```
yarn add artillery-engine-socketio-v3

// OR

npm install artillery-engine-socketio-v3
```

Open a new ticket if any issues come up.

**Comment by dozen1488** [comment]
@ptejada sorry for disturbing, but is it released? I'm using 2.0.19 artillery and `connect` action for socket io.
Nothing mentioned in https://www.artillery.io/docs/reference/engines/socketio docs page (while such option **exists** for WS protocol https://www.artillery.io/docs/reference/engines/websocket)
I'm afraid issue is not colsed.
